src/IO/S3Common.cpp

```diff
@@ -754,16 +754,30 @@ namespace S3
         static constexpr auto OBS = "OBS";
         static constexpr auto OSS = "OSS";
 
-        uri = uri_;
+        uri = Poco::URI(uri_);
+
         storage_name = S3;
 
         if (uri.getHost().empty())
             throw Exception(ErrorCodes::BAD_ARGUMENTS, "Host is empty in S3 URI.");
 
         /// Extract object version ID from query string.
+        bool has_version_id = false;
         for (const auto & [query_key, query_value] : uri.getQueryParameters())
             if (query_key == "versionId")
+            {
                 version_id = query_value;
+                has_version_id = true;
+            }
+
+        /// If uri contains parameter "versionId", the "?" will not be used as wildcards.
+        /// Otherwise it is necessary to encode "?" to avoid deletion during parsing.
```

Comment:
why could it be deleted? why not encode regardless of the has_version_id value?

Reply:
> why could it be deleted? why not encode regardless of the has_version_id value?

Poco::URI will ignore '?' when parsing the path, but if there is a vestionId in the http parameter, '?' can not be used as a wildcard, otherwise it will be ambiguous.